Doc 9195 dev tutorial guides

RayOffiah opened the pull request:
This is the new 'getting started' stuff

* [Couchbase: Getting started](https://rayoffiah.github.io/DOC-9354/server/current/tutorials/couchbase-tutorial-student-records.html)

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress. Also fixed .gitignore to cope with Mac system files and Jetbrains stuff

RayOffiah committed:
Completed the jdk installation bit. Now rolling on to writing an application

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress: added an odd diagram to show how the document database is arranged; added instructions for creating scopes and collections

RayOffiah committed:
Work in progress: starting work on the student java stuff

RayOffiah committed:
Work in progress: finished first example application.

RayOffiah committed:
Work in progress: slight change of tack. We're going to have a separate section as an N1QL introduction.

RayOffiah committed:
Parked after review. Needs more discussion.

RayOffiah committed:
Fixing the "What's next" headings.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress. Adding N1QL search examples

RayOffiah committed:
Tidying up.

RayOffiah committed:
adding-enrolments.adoc added. Moved tutorial to the develop section.

RayOffiah committed:
Interrupted. Stashing work so far

RayOffiah committed:
Continuing with adding-enrolments.adoc

RayOffiah committed:
Continuing with adding-enrolments.adoc

RayOffiah committed:
Fixing callout comments

RayOffiah committed:
Carrying out mods following review

RayOffiah committed:
Fixing install-couchbase-server.adoc

RayOffiah committed:
Adding buckets-scopes-and-collections.adoc

RayOffiah committed:
Fixing images hilarys-record-in-admin-console.png and new-course-collection.png

RayOffiah committed:
Fixing retrieve-documents-using-query-bench-section.adoc and retrieving-documents.adoc

RayOffiah committed:
Fixing retrieve-documents-using-query-bench-section.adoc and retrieving-documents.adoc

RayOffiah committed:
Work in progress

RayOffiah committed:
Menu fix

RayOffiah committed:
Fixing menu and main link

RayOffiah committed:
Work in progress

RayOffiah committed:
Added temporary navigation bar

RayOffiah committed:
Moved the install tabs around to put the local installation first.

RayOffiah committed:
Adding the document design pic to the opening page.

RayOffiah committed:
Work in progress

RayOffiah committed:
Work in progress

RayOffiah committed:
The Python stuff isn't ready yet

RayOffiah committed:
Redoing image to make it a bit clearer

RayOffiah committed:
I've split the menu up a bit to make way for sections in different programming languages

RayOffiah committed:
I've split the menu up a bit to make way for sections in different programming languages

RayOffiah committed:
Rearranging the menu

RayOffiah committed:
Adding the pagination stuff

RayOffiah committed:
Merge branch 'couchbase:release/7.0' into DOC-9195-dev-tutorial-guides

osfameron commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+== Scopes and Collections
+
+There's nothing stopping you from dumping every document into the bucket. But in all but the simplest cases, it's better to provide some kind of separation. between documents of different types. Couchbase has a simple hierarchy model which allows for such separation:
```
extra `.`

RayOffiah replied:
fixed

osfameron commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+
+= Couchbase: Getting started
+
+:description: pass:q[A short tutorial that will guide the developer in downloading and installing Couchbase, then creating a database to store student records.]
```
I don't think `pass` is needed here? iirc the common case is _ that would otherwise italicise text?

simon-dew replied:
Nope. TBH I don't think we should be using markup in the description at all. And please move this to the document header.

```suggestion
:description: A short tutorial that will guide the developer in downloading and installing Couchbase, then creating a database to store student records.
```

RayOffiah replied:
fixed

maria-robobug commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+*collections*:: a collection can contain a set of related documents.
+
+
+NOTE: Remember `scopes` and `collections` are optional: you can insert documents direcly into a `bucket` if you wish.
```
Hmm, if you don't provide a `named` scope and collection, your document will still end up in a scope, it's just called `_default`. Just thought it might be confusing to say it will insert directly into the bucket as it makes me think it won't use scopes and collections entirely.

simon-dew replied:
Likewise don't use fixed width for new terms / definitions, save it for code. We use italics for new terms when necessary.

```suggestion
NOTE: Remember scopes and collections are optional: you can insert documents directly into a bucket if you wish.
```

RayOffiah replied:
Good point, yes

osfameron commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+. Retrieve the `art history` course record
+--
++
+Each method uses a `N1QL` call to retrieve a single record from its respective collection.
```
I don't think we need to format N1QL as code.

RayOffiah replied:
fixed

osfameron commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+{empty}
+<6> Remember that Couchbase doesn't have a native date type, so it's common practice to store the dates as strings.
+
+<7> The `enrollments` inside the student record are stored as an array; since `Json` is the native data format for Couchbase, it's not surprising that each SDK has a host of functions for processing data in Json. In this case, the call `JsonArray.create()` will create an empty list structure.
```
When referring to the data format, JSON in capitals, doesn't need to be formatted as code.

osfameron commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+    .put("date-enrolled", currentDate));
+----
++
+The `course-id` uses the `id` of the course record retrieved from the database. We could, or course, store the whole record in this field. Why *_wouldn't_* we want to do that?
```
italics and bold?

RayOffiah replied:
Yeah, just one'll do.

osfameron commented on modules/tutorials/pages/java-tutorial/first-couchbase-application.adoc:
```diff
@@ -0,0 +1,191 @@
+    public static void main(String[] args) {
+
+        Cluster cluster = Cluster.connect("localhost",
+                "Administrator", "password");    // <.>
```
explicit numbers would be more readable for future?

RayOffiah replied:
Yup

maria-robobug commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+----
+
+
+Couchbase uses a document model which stores each database record as a `json` document. The document can can contain both simple scalar types, and complex types such as nested records and arrays.
```
https://docs.couchbase.com/home/contribute/word-list.html#j

Think we should use `JSON` instead of `json` where possible.

maria-robobug replied:
`The document can can...`
 
 Extra `can`.

osfameron commented on modules/tutorials/pages/java-tutorial/install-couchbase-java-sdk.adoc:
```diff
@@ -0,0 +1,76 @@
+
+You can test the setup is correct by opening a terminal window and changing to the `student` directory. Run the build script to pull in all the dependencies.
+
+[source,bash]
```
`[source,bash]` is for bash scripts, `[source,console]` better for command line transcripts

osfameron commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+Of course, you can also retrieve documents using the JDK. In this section, you're going to use the same `N1QL` queries as part of a small Java application. Let's start with a basic record retrieval:
+
+[source, java]
+----
```
Will all this Java code be moved into SDK and tested?

RayOffiah replied:
At some point, yes.

osfameron commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+
+You may have noticed a difference between the SQL statement we useed in the web console, and the statement used as part of the application:
+
+[source, couchbasequery]
```
isn't this usually `[source,n1ql]` ?

osfameron commented on modules/tutorials/pages/java-tutorial/install-couchbase-java-sdk.adoc:
```diff
@@ -0,0 +1,76 @@
+
+You will need a few things installed on your machine before you begin:
+
+* The Java Software Development Kit (version {minimum-jdk-version})
```
these 2 versions haven't rendered on the sample output

RayOffiah replied:
Good spot!

osfameron commented on modules/tutorials/pages/java-tutorial/first-couchbase-application.adoc:
```diff
@@ -0,0 +1,191 @@
+Working with any database SDK usually involves the following operations:
+
+. Connect to the database
+. Carry out CRUD{crud-footnote} operations
```
Footnote feels a bit heavy. Is there a way to use `<abbr>` tag?
Or just put it in parentheses?

RayOffiah replied:
Could use the <abbrev> tag, but I'm not keen on putting HTML in the Asciidoc.  Might have to see if we can get this:

https://github.com/asciidoctor/asciidoctor/issues/252

moved on a bit. In the meantime, I'll leave it as a footnote.

osfameron commented on modules/tutorials/pages/java-tutorial/first-couchbase-application.adoc:
```diff
@@ -0,0 +1,191 @@
+
+Somewhere in the output, you'll find the line containing the name of the collection you've successfully retrieved:
+
+image::connect-to-cluster-console-output.png[]
```
nice visualisation! I'd make the callout text still bigger though, perhaps remove some of the surrounding text to make space.

osfameron commented on modules/tutorials/pages/java-tutorial/second-couchbase-application.adoc:
```diff
@@ -0,0 +1,124 @@
+.Answer
+[%collapsible]
+=====
+Documents of different types can reside in the same bucket/collection, so it's useful to include a field which can be used to filter documents by type.
```
This jarred a little here - we're explicitly posting these documents to the `course-record-collection` collection, so we don't really need a type?

Do we need an alternative example to contrast better with scope/collection usage?

RayOffiah replied:
Yup, I did think about this, but I think it's more important that folk use the 'type' field to shield themselves from problems later on. I'm the same way with primary fields: you should always have a primary field that isn't related to the data so you can do stuff later on, like change account numbers.

osfameron commented on modules/tutorials/pages/java-tutorial/second-couchbase-application.adoc:
```diff
@@ -0,0 +1,124 @@
+
+== Next steps
+
+So you've created a cluster, a bucket, a scope and two collections. You've also populated your collections with documents. Well, a database isn't much use until we can retrieve information from it, which is what you're going to take a look at in the xref:java-tutorial/retrieving-documents.adoc[next part].
```
Do we now need this explicit link, as we have pagination?

RayOffiah replied:
Not really, but without the link, it's a bit weird. I'll think about rewriting

osfameron commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,152 @@
+
+Type the following query into the query editor field:
+
+[source, couchbasequery]
```
Should be n1ql, which is recognized by the hljs highlighter.
If we want couchbasequery as a synonym (do we?) then should add it with similar to https://github.com/couchbase/docs-ui/pull/119/commits/41af4b125e1faff6a47b1ef75bb06a690e384229

osfameron commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,152 @@
+]
+----
+
+Okay, now try something else: returning the courses with `credit-points` < 200:
```
should `< 200` be code? Or written as "less than" ?

osfameron commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+    public static void main(String[] args) {
+        Cluster cluster = Cluster.connect("localhost",
+                "Administrator", "password");
+
```
here and in one other place, double linespace

osfameron commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+
+Then you'll get a list of the classes in the output.
+
+image::retrieve-courses-cli.png[]
```
this image is a bit too small, would it work as text?

simon-dew commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+// suppress inspection "JsonStandardCompliance" for whole file
+
+:page-pagination:
```
As this is the first page in the sequence, it doesn't make sense to link to the previous page. Need a way to suppress the previous link for this page. Here's a hideous hack to do that.

```suggestion
:page-pagination:

ifdef::basebackend-html[]
++++
<style type="text/css">
  /* No link to previous from this page */
  nav.pagination > span.prev {
    display: none !important;
  }
</style>
++++
endif::[]
```

RayOffiah replied:
Yes, that's probably not the best way to do it. I think we'd be better to add options to the page-pagination attribute so we can specify which links we want:

https://github.com/couchbase/docs-ui/pull/125

If you could have a quick look at this, that would be great. 

I might even suggest it to Dan Allen for inclusion in the Antora build.

osfameron reviewed (commented):
This is looking good!

Some comments inline, mostly minor / stylistic.

Lots of git commits with "Work in Progress". Will these all be squashed into a single commit on merge? Otherwise might be worth rebasing to tidy a little.

Don't love the IntelliJ magic comments, would prefer to minimize editor cruft artefacts, or perhaps add an explanatory comment? Not a big deal though as I guess they're useful :) 

I remember some discussion about Java - will there be other versions for other languages in due course? Also would be great to move those code snippets into SDK space to make sure the code is tested, I think that's already been discussed too?

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+:page-pagination:
```
As this is the last page in the sequence, it doesn't make sense to link to the next page. Need a way to suppress the next link for this page. Here's a hideous hack to do that.

```suggestion
:page-pagination:

ifdef::basebackend-html[]
++++
<style type="text/css">
  /* No link to next from this page */
  nav.pagination > span.next {
    display: none !important;
  }
</style>
++++
endif::[]
```

RayOffiah replied:
see above

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+[horizontal]
+*scopes*:: acts as a parent to a `collection`. You can apply permissions at the scope level, which will restrict access to the collections inside it.
+*collections*:: a collection can contain a set of related documents.
```
No need to bold the definitions in a definition list. Also don't use fixed width (backquotes) for new terms / definitions, if they're not code.

```suggestion
scopes:: acts as a parent to a collection. You can apply permissions at the scope level, which will restrict access to the collections inside it.
collections:: a collection can contain a set of related documents.
```

simon-dew commented on modules/ROOT/nav.adoc:
```diff
@@ -12,6 +13,16 @@
 * xref:getting-started:try-a-query.adoc[Run Your First N1QL Query]
 * xref:getting-started:choose-your-next-steps.adoc[Choose Your Next Steps]
 
+.Tutorial
```
we really need to think about where to put this tutorial and how it interacts with / repeats the Getting started section. I'll think about this some more and add further comments.

RayOffiah replied:
Yes, I've just left it somewhere I can find it. Maybe we need to remove the Getting Started section?

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+= Buckets, Scopes and Collections
+
+:description: pass:q[In this section, you'll learn how to logically partition your data in Couchbase using buckets, scopes and collections.]
```
no need to pass through.
 
```suggestion
:description: In this section, you'll learn how to logically partition your data in Couchbase using buckets, scopes and collections.
```

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+And from this dialog, you can create the `student-bucket`
+
+image::add-student-bucket.png[width="{image-width}"]
```
Please add alt text if possible for screen readers / people who can't view the images, e.g. "Add Data Bucket dialog with student-bucket in the Name box". (Throughout)

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+If you think in database terms, a Couchbase Bucket is analogous to a database: it's the data store where you're going to store and retrieve related information about the students.
+
+
+You can click on the menu:Dashboard[Buckets] link to access the Buckets page, then click on btn:[Add Bucket]
```
Use full stop at end of sentences, even if there is markup at the end of the sentence. (Throughout)

```suggestion
You can click on the menu:Dashboard[Buckets] link to access the Buckets page, then click on btn:[Add Bucket].
```

maria-robobug commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+include::partial$student-database-diagrams.adoc[tags="student-record-erd"]
+
+we can see that our equivalent document-based system could do with a little decomposition:
```
Should be '**W**e' here no?

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+You can click on the menu:Dashboard[Buckets] link to access the Buckets page, then click on btn:[Add Bucket]
+
+And from this dialog, you can create the `student-bucket`
```
Don't rely on the images to tell users what to do in case they are using screen readers or can't view the images. Be specific about what you want them to do. So in this case, say something like:

```suggestion
In this dialog, enter `student-bucket` in the *Name* box.
```

RayOffiah committed:
Replaced ascii diagram blocks with images

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+[abstract]
+{description}
+
+== So what is a Bucket exactly?
```
Use title case for headings (throughout)

```suggestion
== So What is a Bucket Exactly?
```

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+
+There's nothing stopping you from dumping every document into the bucket. But in all but the simplest cases, it's better to provide some kind of separation. between documents of different types. Couchbase has a simple hierarchy model which allows for such separation:
+
+include::partial$student-database-diagrams.adoc[tags="couchbase-hierarchy"]
```
Replace with images (throughout).

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+Now that you understand the basics of scopes and collections, return to your admininstration screen so we can add
+them to your bucket.
+
+=== Adding the student scope
```
Title case for headings throughout. I haven't bothered marking all of these up.

```suggestion
=== Adding the Student Scope
```

simon-dew commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+
+
+
+== The data model [[database-design]]
```
We tend to put section ids on the line before the heading. nb also title case

```suggestion
[[database-design]]
== The Data Model 
```

simon-dew commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+<1> The document is stored in JSON format, which allows for the storage of complex types such as arrays without decomposing them to a second table. In this case, the list of `enrollment` records is stored with the student record.
+
+<2> Each `enrollment` record holds a reference to the course it relates to.
++
```
Personally I think it looks messy having multiple paragraph markup in a callout list but could be just me. I would  have this as a note separate from the callout list.

simon-dew commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+====
+It would be a very bad idea to store the course record with each student:
+
+. It would lead to massive data duplication
```
full stop at end of list items.

simon-dew commented on modules/tutorials/pages/install-couchbase-server.adoc:
```diff
@@ -0,0 +1,92 @@
+include::partial$tutorial-globals.adoc[]
+
+= Install or provision the Couchbase server
+:description: pass:q[A short tutorial that will guide the developer in downloading and installing Couchbase, then creating a database to store our student records.]
```
No pass macro, and tidy up the document header. Please put document title on the first line with the doc attributes below it.

```suggestion
:description: A short tutorial that will guide the developer in downloading and installing Couchbase, then creating a database to store our student records.
```

simon-dew commented on modules/tutorials/pages/install-couchbase-server.adoc:
```diff
@@ -0,0 +1,92 @@
+
+Each Couchbase server can run as its own cluster, or join another cluster to form a multi-node system. For the purposes of this basic tutorial, we're only going to concern ourselves with running as a single cluster.
+
+TIP: You can always find the server's configuration page by pointing your browser at: `http://127.0.0.1:8091`
```
Full stop at end of admonitions, even if markup

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+}
+
+----
+<1> As you'll remember from our xref:java-tutorial/first-couchbase-application.adoc#connecting-to-the-database[first example], the application first has to connect to the Couchbase cluster
```
Full stop at end of list items

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+<5> There are three method calls here:
++
+--
+. Retrieve Hilary's student record
```
Full stop at end of list items

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+
+<2> Then it picks up the correct `Bucket` from the cluster.
+
+<3> Do you remember why we need this `waitUntilReady` function is needed when connecting to the server?
++
+{empty}
++
+[%collapsible]
+.Answer
+====
+{why-waitUntilReady}
+====
```
I like the collapsible answer! Although I can't guarantee it will work in the prod site. I'm not a massive fan of multiple level paragraphs and admonitions within a callout list, but that's just my personal preference, not a style rule

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+
+<7> The `enrollments` inside the student record are stored as an array; since `Json` is the native data format for Couchbase, it's not surprising that each SDK has a host of functions for processing data in Json. In this case, the call `JsonArray.create()` will create an empty list structure.
+
+<8> Now `Json` elements are added to the `enrollments` array. Looking back at the <<enrollments-structure,data structure>> for enrollments, we need store just two items: the course that the enrollment relates to, and the date the student enrolled.
```
Json --> JSON in running text (throughout). And the term JSON doesn't need to be fixed-width.

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,106 @@
+:page-pagination:
+include::partial$tutorial-globals.adoc[]
+
+= Buckets, Scopes and Collections
```
Please put doc title on first line with the doc attributes after it.

simon-dew commented on modules/tutorials/pages/getting-started.adoc:
```diff
@@ -0,0 +1,141 @@
+
+:page-pagination:
+
+= Couchbase: Getting started
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/pages/java-tutorial/adding-enrollments.adoc:
```diff
@@ -0,0 +1,202 @@
+:page-pagination:
+
+= Adding course enrollments
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/pages/java-tutorial/first-couchbase-application.adoc:
```diff
@@ -0,0 +1,191 @@
+:page-pagination:
+
+= Writing your first Couchbase application
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/pages/java-tutorial/install-couchbase-java-sdk.adoc:
```diff
@@ -0,0 +1,76 @@
+:page-pagination:
+
+= Installing the Couchbase Java SDK
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/pages/java-tutorial/install-couchbase-java-sdk.adoc:
```diff
@@ -0,0 +1,76 @@
+
+= Installing the Couchbase Java SDK
+
+:description: pass:q[In this tutorial, you're going to create a skeleton application for interacting with the student database you created in xref:install-couchbase-server.adoc[]]
```
Don't use markup in the description:

```suggestion
:description: In this tutorial, you're going to create a skeleton application for interacting with the student database you created previously.
```

simon-dew commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+
+// suppress inspection "SqlResolve" for whole file
+// suppress inspection "SqlNoDataSourceInspection" for whole file
+= Retrieving documents
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,130 @@
+// suppress inspection "SqlNoDataSourceInspection" for whole file
+= Retrieving documents
+
+:description: pass:q[In this section, you're going to look at two methods of retrieving documents from a collection: you will use the administration console to build simple queries, and also write a short program to retrieve documents matching certain criteria. Both of the methods will introduce `N1QL` (pronounced _nickel_), Couchbase's SQL-based query language.]
+
+[abstract]
+{description}
```
No markup in the description and description should be short — 150 characters or less, I think

```suggestion
:description: In this section, you're going to look at two methods of retrieving documents from a collection.

[abstract]
{description}
You will use the administration console to build simple queries, and also write a short program to retrieve documents matching certain criteria.
Both of the methods will introduce N1QL (pronounced "nickel"), Couchbase's SQL-based query language.
```

simon-dew commented on modules/tutorials/pages/java-tutorial/second-couchbase-application.adoc:
```diff
@@ -0,0 +1,124 @@
+:page-pagination:
+
+= Your second Couchbase application: populating the course details
```
Please put document title on the first line with the doc attributes below it.

RayOffiah replied:
Can't right now. The page-pagination only seems to work reliably as the very first element on the page. Not sure why this is, because attributes don't work like that. I've dropped a note to OpenDevise to see if they can advise.

simon-dew commented on modules/tutorials/pages/python-tutorial/install-couchbase-python-sdk.adoc:
```diff
@@ -0,0 +1,19 @@
+:page-pagination:
+include::partial$tutorial-globals.adoc[]
+= Installing the Couchbase Python SDK
```
Please put document title on the first line with the doc attributes below it.

simon-dew commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,152 @@
+// suppress inspection "JsonStandardCompliance" for whole file
+// suppress inspection "SqlNoDataSourceInspection" for whole file
+// suppress inspection "SqlResolve" for whole file
+= Using the Query Editor
```
Please put document title on the first line with the doc attributes below it.

simon-dew reviewed (changes requested):
Thanks for this @RayOffiah ! I have quite a few changes which are mostly stylistic. In most cases I have just marked these up once and left you to find the other locations where they occur. The biggest one, which I haven't bothered to mark up at all, is perforated prose: please start a new line after every full stop. So every sentence should be in a complete line by itself.

simon-dew commented:
> ventilated prose: please start a new line after every full stop. So every sentence should be in a complete line by itself.

Just to be clear: if a paragraph contains several sentences, there should be no blank line between the sentences. See https://docs.couchbase.com/home/contribute/pages.html#ventilate

RayOffiah committed:
Committing first round of chantes following review

RayOffiah committed:
Committing first round of changes following review

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

RayOffiah committed:
Update modules/tutorials/pages/java-tutorial/retrieving-documents.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Fixups following review

RayOffiah committed:
Fixups following review

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

# Conflicts:
#	modules/tutorials/pages/java-tutorial/retrieving-documents.adoc

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

# Conflicts:
#	modules/tutorials/pages/java-tutorial/retrieving-documents.adoc

RayOffiah committed:
Putting sentences are their own line.

RayOffiah committed:
This file contains the PlantUML definitions used to generate the diagrams.

RayOffiah committed:
Making a couple of pictures bigger

RayOffiah committed:
Correction: there is no default bucket; you mean a default scope.

RayOffiah committed:
Moved examples to Java SDK project.Tidied the headings. Removed the IntelliJ comment tags

RayOffiah committed:
Adding ALT tags to the images.

RayOffiah committed:
Update modules/tutorials/pages/java-tutorial/install-couchbase-java-sdk.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/getting-started.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Fixups from review

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

RayOffiah committed:
Resized images so they centre properly.

RayOffiah committed:
Redoing the console graphics

RayOffiah committed:
Following a review, removing the 'type' field from the tutorial examples.

RayOffiah committed:
Merge branch 'couchbase:release/7.0' into DOC-9195-dev-tutorial-guides

RayOffiah committed:
Changes and tidying up following review.

RayOffiah committed:
Renaming pages.

RayOffiah committed:
DOC-9354 Create new section to onboard someone new to Couchbase
Moving to new location on the menu

RayOffiah committed:
DOC-9354 Create new section to onboard someone new to Couchbase
Adding new page for transactions.

RayOffiah committed:
DOC-9354
Removing the transactions section (overkill) and adding a note pointing folk to the Couchbase Capella documentation.

RayOffiah committed:
DOC-9354
Tidying up.

simon-dew commented on modules/tutorials/pages/install-couchbase-server.adoc:
```diff
@@ -0,0 +1,99 @@
+Once you've downloaded the software, install it on your machine. (The method for installation will depend on your operating system).
+
+The method for running the application will, again, depend on your host operating system.
+In our example, we're using a Mac, so the installation will place then executable app in the standard `Applications` folder:
```
Full stop at end

```suggestion
In our example, we're using a Mac, so the installation will place then executable app in the standard `Applications` folder.
```

simon-dew commented on modules/tutorials/pages/install-couchbase-server.adoc:
```diff
@@ -0,0 +1,99 @@
+[source, bash, subs="+attributes"]
+----
+docker run -t --name db -p 8091-8096:8091-8096 -p 11210-11211:11210-11211 couchbase/server:enterprise-{current-enterprise-version}
+
```
Suggest splitting this over two lines as it overflows the code sample box

```suggestion
docker run -t --name db -p 8091-8096:8091-8096 -p 11210-11211:11210-11211 \
couchbase/server:enterprise-{current-enterprise-version}
```

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,108 @@
+In all but the simplest cases, it's better to provide some kind of separation between documents of different types.
+Couchbase has a simple hierarchy model which allows for such separation:
+
+image::multi-level-data-containment.png[alt="Couchbase data hierarchy"]
```
Just noticed that this image has a disclaimer at the bottom left saying "Confidential and proprietary: do not distribute without Couchbase consent"! It's not a problem for now, but I think we should eventually replace this image, e.g. with something we have generated in UML.

RayOffiah replied:
Good point.  I'll put something together for the follow up release.

simon-dew commented on modules/tutorials/pages/buckets-scopes-and-collections.adoc:
```diff
@@ -0,0 +1,108 @@
+image::add-collection-link.png[alt="Adding a new collection"]
+
+When the collection dialog is displayed, fill in the name of the first collection: `student-record-collection`; then
+press  btn:[save]
```
Capitalization and full stop

```suggestion
press btn:[Save].
```

simon-dew commented on modules/tutorials/partials/student-mvn-directory-structure.adoc:
```diff
@@ -0,0 +1,48 @@
+// # tag::basic[]
```
Just wanted to acknowledge that this is a brilliant way of illustrating a directory structure! 👏

RayOffiah replied:
Thanks :-)

I think nicked the idea from Sarah at OpenDevise. She does something similar for the Asciidoctor documentation.

simon-dew commented on modules/tutorials/partials/tutorial-globals.adoc:
```diff
@@ -0,0 +1,18 @@
+
+:minimum-python-version: 3+
+
+:why-waitUntilReady: pass:q[As discussed earlier in xref:getting-started.adoc[], Couchbase has been designed from the ground up to be scalable and performant, and an important part of this design is the architecture behind the APIs. Most of the APIs are non-blocking, which means that once you call them, your program is free to carry on and do other things while the called function executes.]
```
Broken link

```suggestion
:why-waitUntilReady: pass:q[As discussed earlier in xref:couchbase-tutorial-student-records.adoc[], Couchbase has been designed from the ground up to be scalable and performant, and an important part of this design is the architecture behind the APIs. Most of the APIs are non-blocking, which means that once you call them, your program is free to carry on and do other things while the called function executes.]
```

RayOffiah replied:
Good catch!

simon-dew commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,64 @@
+= Retrieving documents
+:description: In this section, you're going to look at two methods of retrieving documents from a collection: you will use the administration console to build simple queries, and also write a short program to retrieve documents matching certain criteria. Both of the methods will introduce N1QL (pronounced 'nickel'), Couchbase's SQL-based query language
```
Full stop

```suggestion
:description: In this section, you're going to look at two methods of retrieving documents from a collection: you will use the administration console to build simple queries, and also write a short program to retrieve documents matching certain criteria. Both of the methods will introduce N1QL (pronounced 'nickel'), Couchbase's SQL-based query language.
```

simon-dew commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,142 @@
+  }
+]
+----
+<1> The internal Couchbase code for the message
```
Full stop at end for consistency with other callout list items

```suggestion
<1> The internal Couchbase code for the message.
```

simon-dew commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,142 @@
+select META().id, crc.* from `course-record-collection` crc where crc.`credit-points` < 200
+----
+
+The `META()` function call will return any property contained in the document's metadata, including its id
```
Colon at end

```suggestion
The `META()` function call will return any property contained in the document's metadata, including its id:
```

simon-dew commented on modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc:
```diff
@@ -0,0 +1,142 @@
+
+<1> id fields added to the returned document.
+
+You can find a full rundown of the N1QL language here: xref:n1ql:n1ql-language-reference/index.adoc[]
```
Full stop at end

```suggestion
You can find a full rundown of the N1QL language here: xref:n1ql:n1ql-language-reference/index.adoc[].
```

simon-dew commented on modules/tutorials/pages/java-tutorial/retrieving-documents.adoc:
```diff
@@ -0,0 +1,64 @@
+== Next steps
+
+Now you can add and search for records, the next section will consolidate what you've learned so far by demonstrating how to amend existing records by adding enrollment details.
+So when you're ready carry on to xref:java-tutorial/adding-course-enrollments.adoc[]
```
Full stop at end.

```suggestion
So when you're ready carry on to xref:java-tutorial/adding-course-enrollments.adoc[].
```

simon-dew commented on modules/ROOT/nav.adoc:
```diff
@@ -9,7 +10,14 @@
 * xref:getting-started:do-a-quick-install.adoc[Do a Quick Install]
 * xref:getting-started:look-at-the-results.adoc[Explore the Server Configuration]
-* xref:getting-started:try-a-query.adoc[Run Your First N1QL Query]
+* xref:tutorials:couchbase-tutorial-student-records.adoc[]
+  ** xref:tutorials:install-couchbase-server.adoc[]
+  ** xref:tutorials:buckets-scopes-and-collections.adoc[]
+  ** xref:tutorials:java-tutorial/install-couchbase-java-sdk.adoc[]
+  ** xref:tutorials:java-tutorial/creating-the-students-collection.adoc[]
+  ** xref:tutorials:java-tutorial/creating-the-courses-collection.adoc[]
+  ** xref:tutorials:java-tutorial/retrieving-documents.adoc[]
+  ** xref:tutorials:java-tutorial/adding-course-enrollments.adoc[]
```
Reinstate `try-a-query.adoc` for now, at least until we've merged the developer how-to guides. 

```suggestion
* xref:getting-started:try-a-query.adoc[Run Your First N1QL Query]
```

simon-dew commented on modules/ROOT/nav.adoc:
```diff
@@ -155,7 +163,7 @@
 
 .Develop
-* xref:sdk:development-intro.adoc[Developer's Intro]
+//* xref:sdk:development-intro.adoc[Developer's Intro]
```
Let's put the tutorial here for now, at least until we've had a chance to merge the developer how-to guides. I've also given the tutorial a short name for the purposes of the Navigation file. Finally, I've reinstated `development-intro.adoc` for now, although I have renamed it for the purposes of the navigation. We can remove this file for good when we've had a chance to merge the developer how-to guides.

```suggestion
* xref:tutorials:couchbase-tutorial-student-records.adoc[Couchbase Tutorial]
  ** xref:tutorials:install-couchbase-server.adoc[]
  ** xref:tutorials:buckets-scopes-and-collections.adoc[]
  ** xref:tutorials:java-tutorial/install-couchbase-java-sdk.adoc[]
  ** xref:tutorials:java-tutorial/creating-the-students-collection.adoc[]
  ** xref:tutorials:java-tutorial/creating-the-courses-collection.adoc[]
  ** xref:tutorials:java-tutorial/retrieving-documents.adoc[]
  ** xref:tutorials:java-tutorial/adding-course-enrollments.adoc[]
* xref:sdk:development-intro.adoc[Key-Value Operations]
```

simon-dew reviewed (changes requested):
@RayOffiah this is brilliant. Just a few punctuation problems and broken links. The main thing I've asked is that we could put your tutorial at the start of the Develop section for now. We can revisit the Getting Started section and move the tutorial to the start of the nav after we've had a chance to merge the developer how-to guides. Cheers!

RayOffiah committed:
Update modules/tutorials/pages/install-couchbase-server.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/install-couchbase-server.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/buckets-scopes-and-collections.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/pages/java-tutorial/retrieving-documents.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/ROOT/nav.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/ROOT/nav.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/partials/tutorial-globals.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
DOC-9354
Adding new diagram for the Couchbase hierarchy

RayOffiah committed:
Merge remote-tracking branch 'origin/DOC-9195-dev-tutorial-guides' into DOC-9195-dev-tutorial-guides

RayOffiah committed:
Update modules/tutorials/pages/java-tutorial/retrieving-documents.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

RayOffiah committed:
Update modules/tutorials/partials/retrieve-documents-using-query-bench-section.adoc

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

simon-dew reviewed (approved):
LGTM 👍 Sorry for the delay.
